vsintegration/src/FSharp.Editor/Hints/HintService.fs

```diff
@@ -5,59 +5,35 @@ namespace Microsoft.VisualStudio.FSharp.Editor.Hints
-        let rec resolve hintKinds resolvers =
+    let private getHints sourceText parseResults hintKinds symbolUses (symbol: FSharpSymbol) =
+        let rec getHints hintKinds acc =
             match hintKinds with
-            | [] -> resolvers |> Seq.choose id
+            | [] -> acc
             | hintKind :: hintKinds ->
-                match hintKind with
-                | HintKind.TypeHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol -> getTypeHints parseResults symbol |> Some
-                    | _ -> None
-                | HintKind.ReturnTypeHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol -> getReturnTypeHints parseResults symbol |> Some
-                    | _ -> None
-                | HintKind.ParameterNameHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol ->
-                        getHintsForMemberOrFunctionOrValue sourceText parseResults symbol |> Some
-                    | :? FSharpUnionCase as symbol -> getHintsForUnionCase parseResults symbol |> Some
-                    | _ -> None
-                // we'll be adding other stuff gradually here
-                :: resolvers
-                |> resolve hintKinds
-
-        resolve hintKinds []
-
-    let private getHintsForSymbol (sourceText: SourceText) parseResults hintKinds (symbol: FSharpSymbol, symbolUses: FSharpSymbolUse seq) =
-        symbol
-        |> getHintResolvers sourceText parseResults hintKinds
-        |> Seq.collect (fun resolve -> resolve symbolUses)
+                match (hintKind, symbol) with
+                | HintKind.TypeHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses |> Seq.collect (InlineTypeHints(parseResults, symbol)).getHints
+                | HintKind.ReturnTypeHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses |> Seq.collect (InlineReturnTypeHints(parseResults, symbol).getHints)
+                | HintKind.ParameterNameHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses
+                    |> Seq.collect (InlineParameterNameHints(parseResults).getHintsForMemberOrFunctionOrValue sourceText symbol)
+                | HintKind.ParameterNameHint, (:? FSharpUnionCase as symbol) ->
+                    symbolUses
+                    |> Seq.collect (InlineParameterNameHints(parseResults).getHintsForUnionCase symbol)
+                | _ -> []
+                :: acc
+                |> getHints hintKinds
+
+        getHints (hintKinds |> Set.toList) []
+
+    let private getHintsForSymbol (sourceText: SourceText) parseResults hintKinds (symbol, symbolUses) =
+        symbol |> getHints sourceText parseResults hintKinds symbolUses |> Seq.concat
```

Comment:
Can we please have less piping here? It doesnt really bring much here, just creating a bunch of unnecessary partially applied functions.

Something along the lines of
```suggestion
        let hints = getHints sourceText parseResults hintKinds symbolUses symbol 
        Seq.concat hints
```

Reply:
It's partly remainders from previous versions of the code, partly personal preference. Don't have a strong stance here though, updated.